PMNS_NSIs.cxx

```diff
@@ -0,0 +1,409 @@
+///
+/// Set eta_mumu parameter
+///
+/// This will check if value is changing to keep track of whether
+/// the eigensystem needs to be recomputed.
+///
+/// @param a - The value of the real parameter eta_mumu
+///
+void PMNS_NSIs::SetEta_mumu(double a){ SetEta(1,1, a, 0); }
+
+//......................................................................
+///
+/// Set eta_tautau parameter
+///
+/// This will check if value is changing to keep track of whether
+/// the eigensystem needs to be recomputed.
+///
+/// @param a - The value of the real parameter eta_tautau
+///
+void PMNS_NSIs::SetEta_tautau(double a){ SetEta(2,2, a, 0); }
+
+//......................................................................
+
+void PMNS_NSIs::SetEta_emu(double a, double phi){ SetEta(0,1, a, phi); }
+
+//......................................................................
+///
+/// Set eta_mumu parameter
+///
+/// This will check if value is changing to keep track of whether
+/// the eigensystem needs to be recomputed.
+///
+/// @param a - The value of the real parameter eta_mumu
+///
+void PMNS_NSIs::SetEta_mutau(double a, double phi){ SetEta(1,2, a, phi); }
+
+//......................................................................
+///
+/// Set eta_tautau parameter
+///
+/// This will check if value is changing to keep track of whether
+/// the eigensystem needs to be recomputed.
+///
+/// @param a - The value of the real parameter eta_tautau
+///
+void PMNS_NSIs::SetEta_etau(double a, double phi){ SetEta(0,2, a, phi); }
+
+
+////Set lightest neutrino mass
+void PMNS_NSIs::SetM(double m){ SetLowestMass(m);}
+
+double PMNS_NSIs::GetM(){ return fM;}
+/// Set eps_mutau parameter
+///
+/// This will check if value is changing to keep track of whether
+/// the eigensystem needs to be recomputed.
+///
+/// @param a   - The absolute value of the parameter eps_mutau
+///
+/// Build the full Hamiltonian in matter.
+///
+/// Here we divide the mass squared matrix Hms by the 2E
+/// to obtain the vacuum Hamiltonian in eV. Then, the matter
+/// potential is added to the electron component.
+///
+void PMNS_NSIs::UpdateHam()
+{
+
+  double lv = 2 * kGeV2eV*fEnergy;     // 2E in eV 
+
+  double kr2GNe = kK2*M_SQRT2*kGf;
+  kr2GNe *= fPath.density * fPath.zoa; // Matter potential in eV
+
+  // Finish building Hamiltonian in matter with dimension of eV
+  double  s12, s23, s13, c12, c23, c13;
+  complexD idelta(0.0, fDelta[0][2]);
+  if(fIsNuBar) idelta = conj(idelta);
+
+  s12 = sin(fTheta[0][1]);  s23 = sin(fTheta[1][2]);  s13 = sin(fTheta[0][2]);
+  c12 = cos(fTheta[0][1]);  c23 = cos(fTheta[1][2]);  c13 = cos(fTheta[0][2]);
+  complexD U[3][3];
+  U[0][0] =  c12*c13;
+  U[0][1] =  s12*c13;
+  U[0][2] =  s13*exp(-idelta);
+
+  U[1][0] = -s12*c23-c12*s23*s13*exp(idelta);
+  U[1][1] =  c12*c23-s12*s23*s13*exp(idelta);
+  U[1][2] =  s23*c13;
+
+  U[2][0] =  s12*s23-c12*c23*s13*exp(idelta);
+  U[2][1] = -c12*s23-s12*c23*s13*exp(idelta);
+  U[2][2] =  c23*c13;
```

Comment:
This is basically the `fEvec` in vacuum. Can we not get this from existing code already?

Reply:
Used fEvec.